src/strategies/properties/attribute.sml

```diff
@@ -0,0 +1,106 @@
+import "util.parser";
+import "util.multiset";
+import "util.type";
+
+signature ATTRIBUTE =
+sig
+  structure M : MULTISET;
+  type T;
+  val fromType : Type.T -> T;
+  val fromHoles : Type.T M.multiset -> T;
+  val fromContent : Type.T -> T;
+  val fromFeature : string -> T;
+  val read : string -> T list;
+  val toString : T -> string;
+  val getType : T list -> Type.T option
+end
+
+
+structure Attribute : ATTRIBUTE =
+struct
+
+  structure M = Multiset(struct
+                         type t = Type.T;
+                         val compare = Type.compare;
+                         val fmt = Type.toString;
+                         end);
+
+  datatype T = IsOfType of Type.T
+             | Holes of Type.T M.multiset
+             | Tokens of string list
+             | Content of Type.T
+             | Occurrences of int
+             | Feature of string;
+
+  fun fromType t = IsOfType t;
+  fun fromHoles H = Holes H;
+  fun fromTokens L = Tokens L;
+  fun fromContent c = Content c;
+  fun fromFeature s = Feature s;
+
+  fun decomposeAttribute a =
+      let val (x,_,y) = (Parser.breakOn ":=" a);
+      in (Parser.stripSpaces x,  Parser.stripSpaces y)
+      end;
+
+  fun intFromString s =
+      case Int.fromString s of
+              SOME n => n
+            | NONE => if s = "#t" then ~1 else if s = "sqrt(#t)" then ~2 else if s = "log(#t)" then ~3
+                        else (print ("bad numerical expression: " ^ s); raise Parser.ParseError);
+
+  fun makeHoleList [] = []
+    | makeHoleList (a::L) =
+        let val (sx,_,sy) = Parser.breakOn "=>" a;
+            val y = Parser.stripSpaces sy;
+            val x' = Type.fromString sx;
+            val y' = intFromString y;
+        in (x',y') :: makeHoleList L
+        end;
+
+  fun holeMultisetFromList L = M.fromPairList (makeHoleList L);
+
+(* holeMultisetFromString takes a string with square brackets with
+    type/multiplicities pairs separated by dots [int => 1. real => 4. set => 1]*)
+  fun holeMultisetFromString s =
+      let val L = Parser.splitStrip "." (Parser.removeSquareBrackets s)
+      in holeMultisetFromList L
+      end
+
+  fun tokenListFromString s = Parser.splitStrip "." (Parser.removeSquareBrackets s)
+
+  fun holePairToString (t,n) = (Type.toString t) ^ " => " ^ (Int.toString n)
+
+  fun attributeFromPair (x,y) =
+      if y = "" then Feature x
+      else if x = "type" then IsOfType (Type.fromString y)
+      else if x = "content" then Content (Type.fromString y)
+      else if x = "holes" then Holes (holeMultisetFromString y)
+      else if x = "tokens" then Tokens (tokenListFromString y)
+      else if x = "occurrences" then Occurrences (intFromString y)
+      else (print ("error parsing attribute: " ^ x ^ " " ^ y); raise Parser.ParseError)
+
+  fun read s =
```

Comment:
This is slightly strange, for a two reasons.

1. Why is it not called fromString?
2. Why does it read the whole attribute**s** input rather than a single attribute?

I realise the second question partly answers the first, but the second should be considered in reference to the `property.sml` file.

Reply:
Note that by making it individually focused, you end up defining `val fromString = attributeFromPair o decomposeAttribute` and moving the `splitStrip` down to `property.sml`'s `findAttributes`.

Reply:
ok. I made a fromString function, focused on individual attributes, and then the readAttributes function in property.sml deals with multiple.